### PR TITLE
#579: Make content script injector aware of content script itself

### PR DESCRIPTION
**Pull request by fregante**
- Part of #579 
- Replaces https://github.com/pixiebrix/pixiebrix-extension/pull/654

To avoid sending huge PRs, I'm splitting this work over a few PRs. In this one I'm trying to preserve the general behavior while setting it up for further potential improvements. This PR:

- [x] Does away with repeated polling
- [x] Avoids duplicate injection (tabs with permanent permissions are just handled by `webext-dynamic-content-script`)
- [X] Creates a single point to "ready the content script", whether it's just waiting for the `webext-dynamic-content-script`, attempting to inject on a forbidden page, or manually injecting it

---

Ready except:

- [x] Pick better `pbReady` message
- [x] Pick better `pbReady` global

General direction for future PRs:

- the lifters/messengers should just attempt to send the message first and then await/inject the content script automatically

We'll see how practical that will be though. The intent is avoid/limit calling the function itself and also to review the injection and communication overall.

### CODE REVIEW

File: src/background/util.ts [hunk 1]
```diff
@@ -66,46 +42,126 @@
...
+      // This imitates the new chrome.scripting API by wrapping a function in a IIFE
+      code: `(${() => ({
+        url: location.href,
+        installed: Symbol.for("pixiebrix-content-script") in window,
+        ready: (window as any).pbReady,
+      })})()`,
+      frameId: target.frameId,
+    })
+    .then(([state]: [TargetState]) => state)
+    .catch(() => undefined);
```
Comment (twschiller) [inline src/background/util.ts:81]: Codacy has a fix for the issue: [Do not use useless `undefined`.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7684115)
```suggestion
    .catch(() => {});
```

File: src/contentScript/context.ts [hunk 2]
```diff
@@ -45,6 +45,7 @@
 
 export function markReady(): void {
   ready = true;
+  (window as any).pbReady = true;
```
Comment (twschiller) [inline src/contentScript/context.ts:48]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7684115)
```suggestion
  (window as unknown).pbReady = true;
```

File: src/background/util.ts [hunk 3]
```diff
@@ -66,46 +42,126 @@
...
+    );
+  }
+
+  return browser.tabs
+    .executeScript(target.tabId, {
+      // This imitates the new chrome.scripting API by wrapping a function in a IIFE
+      code: `(${() => ({
+        url: location.href,
+        installed: Symbol.for("pixiebrix-content-script") in window,
+        ready: (window as any).pbReady,
```
Comment (twschiller) [inline src/background/util.ts:76]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7684115)
```suggestion
        ready: (window as unknown).pbReady,
```

File: src/contentScript/context.ts (hunk 2 above)
Comment (twschiller) [inline src/contentScript/context.ts:48]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7684115)
```suggestion
  (window as unknown).pbReady = true;
```

File: src/background/util.ts (hunk 1 above)
Comment (twschiller) [inline src/background/util.ts:81]: Codacy has a fix for the issue: [Do not use useless `undefined`.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7684115)
```suggestion
    .catch(() => {});
```

File: src/background/util.ts (hunk 3 above)
Comment (twschiller) [inline src/background/util.ts:76]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7684115)
```suggestion
        ready: (window as unknown).pbReady,
```

File: src/background/util.ts [hunk 4]
```diff
@@ -66,46 +42,126 @@ export async function waitReady(
...
+}
+
+interface TargetState {
+  url: string;
+  installed: boolean;
+  ready: boolean;
+}
+
+/** Fetches the URL from a tab/frame. Return value is empty if it's unknowable (likely because we don't have permission to access it) */
+export async function getTargetState(
```
[the remark this answers is not in the document]
Comment (fregante) [reply src/background/util.ts:44]: Actually, I'll make this throw. This way the reason why it failed will bubble up

File: src/background/devtools/internal.ts [hunk 5]
```diff
@@ -268,18 +269,29 @@ export function registerPort(tabId: TabId, port: Runtime.Port): void {
...
+    await ensureContentScript({ tabId, frameId });
+  } catch (error) {
+    if (
+      isErrorObject(error) &&
+      error.message.startsWith("No access to frame")
+    ) {
       console.debug(
-        `Skipping injectDevtoolsContentScript because no activeTab permissions for tab: ${tabId}`
+        `Skipping injectDevtoolsContentScript because no activeTab permissions`,
+        { tabId, frameId, url }
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/background/devtools/internal.ts:293]: You mean if the devtools are connected to another frame on the tab? (We haven't really though through the page editor functionality/handling for frames)

Please leave this note in the code comments, as it's important context for future work

File: src/contentScript/context.ts (hunk 2 above)
Comment (twschiller) [reply src/contentScript/context.ts:45]: Where is this explained?

In general, would be good to formally write the types for the window interface in a file so there's one place to see what bookkeeping we've added in the PixieBrix codebase. https://mariusschulz.com/blog/declaring-global-variables-in-typescript#augmenting-the-window-interface.

File: src/background/devtools/internal.ts [hunk 6]
```diff
@@ -308,7 +323,7 @@ if (isBackgroundPage()) {
   });
 
   browser.webNavigation.onDOMContentLoaded.addListener((details) => {
-    void injectTemporaryAccess(details);
     emitDevtools("DOMContentLoaded", details);
```
Comment (twschiller) [inline src/background/devtools/internal.ts:326]: Could you add a comment here on why this is the correct sequence of calls? It seems weird that we'd want the devtools to handle DOMContentLoaded before it has access

File: src/background/util.ts [hunk 7]
```diff
@@ -66,46 +42,126 @@ export async function waitReady(
...
+      frameId: target.frameId,
+    })
+    .then(([state]: [TargetState]) => state)
+    .catch(() => undefined);
+}
+
+export async function onReadyNotification(signal: AbortSignal): Promise<true> {
+  return new Promise((resolve) => {
+    const onMessage = (message: unknown) => {
+      if (message === "pbReady") {
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/background/util.ts:66]: For consistency, I have the preference for using `{type: PIXIEBRIX_READY}`. Even if we keep the string, I have the preference for using a shared constant for sending/handling the message

Maybe throw it in background/protocol for now? As part of the messaging overhaul in the future, we'll be figuring how where/how to refactor the action constants/payload typings

File: src/contentScript/executor.ts [hunk 8]
```diff
@@ -109,6 +109,7 @@ export async function whoAmI(): Promise<Runtime.MessageSender> {
 }
 
 export async function notifyReady(): Promise<void> {
+  void browser.runtime.sendMessage("pbReady");
```
Comment (twschiller) [inline src/contentScript/executor.ts:112]: It would be helpful to comment here why we have to send two "ready" messages - i.e., what role each is playing

File: src/devTools/context.ts [hunk 9]
```diff
@@ -126,18 +126,14 @@ async function runInMillis<TResult>(
...
     throw new PermissionsError(`No access to URL: ${url}`);
   }
   console.debug(`connectToFrame: ensuring contentScript for ${url}`);
-  await injectScript(port, { file: "contentScript.js" });
-
-  console.debug(
-    `connectToFrame: waiting for contentScript to be ready for ${url}`
-  );
-  await waitReady(port, { maxWaitMillis: 4000 });
+  await pTimeout(ensureScript(port), 4000, "contentScript not ready in 4s");
```
Comment (twschiller) [inline src/devTools/context.ts:136]: I'm curious why we're not receiving lint warnings that the 4000 should be `4_000`. Is that rule disabled?

**Comment by twschiller** [comment]
Overall this looks good. With complicated items like this, I have a strong preference for ensuring there's inline documentation of considerations, which sequencing is important, etc.

In general, reviewing this PR has affirmed that we need to have testing infrastructure in place prior to re-architecting the messaging infrastructure

Let me know when this is ready for testing and I'll help with manual QA. Please merge/rebase this PR, so we know what the impact is

File: src/devTools/context.ts (hunk 9 above)
Comment (fregante) [reply src/devTools/context.ts:136]: 4000 is accepted by that rule because it's short enough to not be hard to read. 40000 will show an error

firefox

File: src/background/devtools/internal.ts (hunk 6 above)
Comment (fregante) [reply src/background/devtools/internal.ts:326]: `injectTemporaryAccess` was not awaited, so I'd ask you the same question 😃 
I moved it lower simply because the injector was logging before and after emitDevtools (because some code was execute before, and some after)

File: src/background/devtools/internal.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (fregante) [reply src/background/devtools/internal.ts:293]: I was talking about its behavior in general. I haven't really looked into how/why it deals with frames here other than ensuring that it throws errors correctly.

Will add the comment 👍

File: src/contentScript/context.ts (hunk 2 above)
Comment (fregante) [reply src/contentScript/context.ts:45]: https://github.com/pixiebrix/pixiebrix-extension/pull/661#discussion_r661596020

File: src/background/util.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (fregante) [reply src/background/util.ts:66]: `background/protocol` has a lot of dependencies and code, I'll probably use `messaging/constants.ts` for now. Related:

- https://github.com/pixiebrix/pixiebrix-extension/issues/669

**Comment by fregante** [comment]
Review followed. If you see 🚀 on your review messages it's because I followed them. I do this only when I **also** leave a comment, because if I "resolve" the conversation you might miss it.

File: package.json [hunk 10]
```diff
@@ -220,7 +220,7 @@
       "^.+\\.txt$": "jest-raw-loader"
     },
     "transformIgnorePatterns": [
-      "node_modules/(?!@cfworker|webext-detect-page|idb|webext-patterns|webext-polyfill-kinda|p-timeout|p-defer)"
+      "node_modules/(?!@cfworker|webext-detect-page|idb|webext-patterns|webext-polyfill-kinda|webext-additional-permissions|p-timeout|p-defer)"
```
Comment (fregante) [inline package.json:223]: I'll look into enabling native ESM support in Jest at some point 😅  https://stackoverflow.com/a/61653104

File: src/background/devtools/internal.ts (hunk 6 above)
Comment (twschiller) [reply src/background/devtools/internal.ts:333]: I think the order doesn't matter. As with everything involving communication, there's a lot of layers and it's hard to follow...

Eventually it gets to here: http://github.com/pixiebrix/pixiebrix-extension/blob/bc8c32d237056630d459b60bb445da5d373500b0/src/devTools/context.ts#L213-L213

Which calls http://github.com/pixiebrix/pixiebrix-extension/blob/bc8c32d237056630d459b60bb445da5d373500b0/src/devTools/context.ts#L132-L132

Which I think ends up handling the race condition through a combination of injectScript and waitReady

Comment (fregante) [reply src/background/devtools/internal.ts:333]: `waitReady` only exists for `executeInTarget` -> `MESSAGE_RUN_BLOCK_TARGET` now.

I haven't looked into `emitDevtools("DOMContentLoaded")` specifically, but if you expect it to run when the CS is ready, then it should await `attemptTemporaryAccess`.

**Review by twschiller — APPROVED** [review]
OK to merge - will address any connection corner cases as they arise
